From 506db170d8297f9bf371d26f2b1cf565ff266b8a Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Wed, 16 Jul 2014 20:42:40 -0400
Subject: [md] dm-mpath: do not activate failed paths

Message-id: <1405543364-20901-37-git-send-email-snitzer@redhat.com>
Patchwork-id: 85922
O-Subject: [RHEL7.1 PATCH 36/40] dm-mpath: do not activate failed paths
Bugzilla: 1117872
RH-Acked-by: Mikulas Patocka <mpatocka@redhat.com>
RH-Acked-by: Heinz Mauelshagen <heinzm@redhat.com>

BZ: 1117872

Upstream commit 3a0175096423856d145e210527271c9eec188a5f
Author: Hannes Reinecke <hare@suse.de>
Date:   Fri Feb 28 15:33:49 2014 +0100

    dm-mpath: do not activate failed paths

    activate_path() is run without a lock, so the path might be
    set to failed before activate_path() had a chance to run.
    This patch add a check for ->active in activate_path() to
    avoid unnecessary overhead by calling functions which are known
    to be failing.

    Signed-off-by: Hannes Reinecke <hare@suse.de>
    Reviewed-by: Jun'ichi Nomura <j-nomura@ce.jp.nec.com>
    Signed-off-by: Mike Snitzer <snitzer@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/dm-mpath.c b/drivers/md/dm-mpath.c
index de2f97f..7d6ba33 100644
--- a/drivers/md/dm-mpath.c
+++ b/drivers/md/dm-mpath.c
@@ -1195,8 +1195,11 @@ static void activate_path(struct work_struct *work)
  struct pgpath *pgpath =
   container_of(work, struct pgpath, activate_path.work);
 
- scsi_dh_activate(bdev_get_queue(pgpath->path.dev->bdev),
-    pg_init_done, pgpath);
+ if (pgpath->is_active)
+  scsi_dh_activate(bdev_get_queue(pgpath->path.dev->bdev),
+     pg_init_done, pgpath);
+ else
+  pg_init_done(pgpath, SCSI_DH_DEV_OFFLINED);
 }
 
 static int noretry_error(int error)
-- 
1.7.1